python/ray/data/dataset.py

```diff
@@ -655,6 +655,55 @@ def sort(self,
         """
         return Dataset(sort_impl(self._blocks, key, descending))
 
+    def zip(self, other: "Dataset[U]") -> "Dataset[(T, U)]":
+        """Zip this dataset with the elements of another.
+
+        The datasets must have identical num rows, block types, and block sizes
+        (e.g., one was produced from a ``.map()`` of another).
+
+        Time complexity: O(dataset size / parallelism)
+
+        Args:
+            other: The dataset to zip with on the right hand side.
+
+        Examples:
+            >>> ds1 = ray.data.range(5)
+            >>> ds2 = ray.data.range(5).map(lambda x: x + 1)
+            >>> ds.take()
+            [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]
+
+        Returns:
+            A Dataset with (k, v) pairs (or concatenated Arrow schema) where k
+            comes from the first dataset and v comes from the second.
+        """
+
+        blocks1 = self.get_blocks()
+        blocks2 = other.get_blocks()
+
+        if len(blocks1) != len(blocks2):
```

Comment:
should we add a TODO to support zip two dataset with different number of blocks, but the number rows are identical?